Sources/OAuthenticator/Services/GoogleAPI.swift

```diff
@@ -0,0 +1,221 @@
+                
+                let response = try JSONDecoder().decode(GoogleAPI.OAuthResponse.self, from: data)
+                return response.login
+            }
+            catch let decodingError as DecodingError {
+                os_log(.fault, "Reponse from AuthenticationProvider is not conformed to provided response format. %s", decodingError.failureReason ?? decodingError.localizedDescription)
+                throw decodingError
+            }
+		}
+	}
+
+    /// Token Refreshing
+    /// - Create the request that will refresh the access token from the information in the Login
+    ///
+    /// - Parameters:
+    ///   - login: The current Login object containing the refresh token
+    ///   - appCredentials: The Application credentials
+    /// - Returns: The URLRequest to refresh the access token
+    static func authenticationRefreshRequest(login: Login, appCredentials: AppCredentials) throws -> URLRequest {
+        guard let refreshToken = login.refreshToken,
+              !refreshToken.value.isEmpty else { throw AuthenticatorError.missingRefreshToken }
+
+        var urlBuilder = URLComponents()
+
+        urlBuilder.scheme = GoogleAPI.scheme
+        urlBuilder.host = GoogleAPI.tokenHost
+        urlBuilder.path = GoogleAPI.tokenPath
+        urlBuilder.queryItems = [
+            URLQueryItem(name: GoogleAPI.clientIDKey, value: appCredentials.clientId),
+            URLQueryItem(name: GoogleAPI.refreshTokenKey, value: refreshToken.value),
+            URLQueryItem(name: GoogleAPI.grantTypeKey, value: GoogleAPI.grantTypeRefreshToken),
+        ]
+        
+        guard let url = urlBuilder.url else {
+            throw AuthenticatorError.missingTokenURL
+        }
+
+        var request = URLRequest(url: url)
+        request.httpMethod = "POST"
+        request.setValue("application/json", forHTTPHeaderField: "Accept")
+
+        return request
+    }
+    
+	static func refreshProvider(with parameters: AppCredentials) -> TokenHandling.RefreshProvider {
+		return { login, appCredentials, urlLoader in
+            let request = try authenticationRefreshRequest(login: login, appCredentials: appCredentials)
+            let (data, _) = try await urlLoader(request)
+
+            let jsonString = String(data: data, encoding: .utf8) ?? ""
+            os_log(.debug, "[Authentication Refresh JSON Result] %s", jsonString)
+            
+            do {
+                let response = try JSONDecoder().decode(GoogleAPI.OAuthResponse.self, from: data)
+                return response.login
+            }
+            catch let decodingError as DecodingError {
+                os_log(.fault, "Non-conformant response from AuthenticationProvider: %s", decodingError.failureReason ?? decodingError.localizedDescription)
+                throw decodingError
+            }
+        }
+	}
+}
+
+public extension URL {
+    /// Add the ability to retrieve the scope query item from the URL
+    var grantedScope: String {
```

Comment:
This is an interesting property! There are a few others in URL+Code.swift. Should we maybe rename that file and move more in there?

Reply:
Good idea. I have move this code into a renamed `URL+QueryParams` file.